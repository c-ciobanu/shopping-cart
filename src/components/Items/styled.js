import styled from "styled-components";

import { ReactComponent as TrashIcon } from "assets/svg/trash.svg";

export const StyledList = styled.ul`
	list-style-type: none;
	padding: 0;
`;

export const StyledListItem = styled.li`
	display: grid;
	grid-template-columns: 1fr auto;
	grid-column-gap: 10px;
	align-items: center;
	padding: 10px 0;
	border-bottom: 1px solid #000;

	&:last-child {
		border-bottom: 0;
	}
`;

export const StyledTrashIcon = styled(TrashIcon)`
	height: 16px;
	width: 16px;
	cursor: pointer;
`;
