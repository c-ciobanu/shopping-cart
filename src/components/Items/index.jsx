import React from "react";
import PropTypes from "prop-types";

import Checkbox from "components/Checkbox";

import { StyledList, StyledListItem, StyledTrashIcon } from "./styled";

function Items({ options, onChange, onDelete }) {
	const handleChange = (item, newVal) => {
		onChange(item, newVal);
	};

	return (
		<StyledList>
			{options.map((option) => (
				<StyledListItem key={`item-${option.name}`}>
					<Checkbox name={option.name} label={option.name} onChange={(e) => handleChange(option, e.target.checked)} />

					<StyledTrashIcon onClick={() => onDelete(option)} data-testid={`remove-item-${option.name}`} />
				</StyledListItem>
			))}
		</StyledList>
	);
}

Items.propTypes = {
	options: PropTypes.arrayOf(
		PropTypes.shape({
			name: PropTypes.string.isRequired,
			bought: PropTypes.bool.isRequired
		})
	).isRequired,
	onChange: PropTypes.func.isRequired,
	onDelete: PropTypes.func.isRequired
};

export default Items;
