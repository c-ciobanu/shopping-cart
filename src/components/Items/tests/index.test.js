import React from "react";
import { render, fireEvent, getByTestId } from "@testing-library/react";

import Items from "../index";

describe("<Items />", () => {
	const options = [
		{ name: "Option 1", bought: false },
		{ name: "Option 2", bought: false }
	];

	it("Expect to not log errors in console", () => {
		const spy = jest.spyOn(global.console, "error");
		render(<Items options={[]} onChange={() => {}} onDelete={() => {}} />);
		expect(spy).not.toHaveBeenCalled();
	});

	it("Should correctly call onChange callback on checkbox click", () => {
		const onChangeMock = jest.fn();
		const { getByText } = render(<Items options={options} onChange={onChangeMock} onDelete={() => {}} />);

		fireEvent.click(getByText(options[0].name));

		expect(onChangeMock).toHaveBeenCalled();
		expect(onChangeMock).toHaveBeenCalledWith(options[0], true);
	});

	it("Should correctly call onDelete callback on trash icon click", () => {
		const onDeleteMock = jest.fn();
		const { getByTestId } = render(<Items options={options} onChange={() => {}} onDelete={onDeleteMock} />);

		fireEvent.click(getByTestId(`remove-item-${options[0].name}`));

		expect(onDeleteMock).toHaveBeenCalled();
		expect(onDeleteMock).toHaveBeenCalledWith(options[0]);
	});

	it("Should render and match the snapshot", () => {
		const {
			container: { firstChild }
		} = render(<Items options={[]} onChange={() => {}} onDelete={() => {}} />);
		expect(firstChild).toMatchSnapshot();
	});
});
