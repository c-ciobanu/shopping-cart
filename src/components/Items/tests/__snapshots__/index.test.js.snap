// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Items /> Should render and match the snapshot 1`] = `
<ul
  class="sc-AykKE cRKDsa"
/>
`;
